File: 15.4/ex154.py
from abc import ABC, abstractmethod

class IMark(ABC):
    @abstractmethod
    def input(self):
        pass

    @abstractmethod
    def display(self):
        pass

class StudentAptech(IMark):
    def __init__(self, StuId='', StuName='', gender='', birthday='', nativePlace=''):
        self.StuId = StuId
        self.StuName = StuName
        self.gender = gender
        self.birthday = birthday
        self.nativePlace = nativePlace

    def input(self):
        self.StuId = input("Enter student ID: ")
        self.StuName = input("Enter student name: ")
        self.gender = input("Enter gender: ")
        self.birthday = input("Enter birthday: ")
        self.nativePlace = input("Enter native place: ")

    def display(self):
        print(f"Student ID: {self.StuId}")
        print(f"Student name: {self.StuName}")
        print(f"Gender: {self.gender}")
        print(f"Birthday: {self.birthday}")
        print(f"Native place: {self.nativePlace}")

class StudentMark(IMark):
    def __init__(self, StuId='', className='', subjectName='', semester=0, mark=0.0):
        self.StuId = StuId
        self.className = className
        self.subjectName = subjectName
        self.semester = semester
        self.mark = mark

    def input(self):
        self.StuId = input("Enter student ID: ")
        self.className = input("Enter class name: ")
        self.subjectName = input("Enter subject name: ")
        self.semester = int(input("Enter semester: "))
        self.mark = float(input("Enter mark: "))

    def display(self):
        print(f"Student ID: {self.StuId}")
        print(f"Class name: {self.className}")
        print(f"Subject name: {self.subjectName}")
        print(f"Semester: {self.semester}")
        print(f"Mark: {self.mark}")

class StudentMarkTotal(StudentMark):
    def __init__(self, StuId='', className='', subjectName='', semester=0, mark=0.0, totalExamSubject=0, averageMark=0.0):
        super().__init__(StuId, className, subjectName, semester, mark)
        self.totalExamSubject = totalExamSubject
        self.averageMark = averageMark

    def getTotalExamSubject(self, list):
        self.totalExamSubject = len(list)

    def calculateAverageMark(self, list):
        sum = 0
        for s in list:
            sum += s.mark
        self.averageMark = sum / len(list)

def main():
    students = []
    marks = []

    while True:
        print("1. Enter information for n Aptech students.")
        print("2. Enter m exam scores for these students.")
        print("3. Sort students by name and display.")
        print("4. Find exam score information by student id.")
        print("5. Exit")
        choice = int(input("Choose an option: "))

        if choice == 1:
            n = int(input("Enter the number of students: "))
            for i in range(n):
                print(f"Enter information for student {i + 1}:")
                student = StudentAptech()
                student.input()
                students.append(student)

        elif choice == 2:
            m = int(input("Enter the number of exam scores: "))
            for i in range(m):
                print(f"Enter exam score for student {i + 1}:")
                mark = StudentMarkTotal()
                mark.input()

                student_found = False
                for student in students:
                    if student.StuId == mark.StuId:
                        student_found = True
                        break

                if student_found:
                    marks.append(mark)
                else:
                    print("Student ID not found in the list of students. Please try again.")

        elif choice == 3:
            students.sort(key=lambda x: x.StuName)
            for student in students:
                student.display()

        elif choice == 4:
            search_id = input("Enter the student ID to search: ")
            found = False
            for mark in marks:
                if mark.StuId == search_id:
                    mark.display()
                    found = True

            if not found:
                print("No exam score information found for the given student ID.")

        elif choice == 5:
            print("Exiting.")
            break

        else:
            print("Invalid choice. Please try again.")

if __name__ == "__main__":
    main()

File: 15.4/ex154.java
import java.util.Scanner;

public interface ic {
    public void input();
    public void display();
}

public class StudentAptech implements ic {
    private String StuId;
    private String StuName;
    private String gender;
    private String birthday;
    private String nativePlace;

    public StudentAptech() {
    }

    public StudentAptech(String StuId, String StuName, String gender, String birthday, String nativePlace) {
        this.StuId = StuId;
        this.StuName = StuName;
        this.gender = gender;
        this.birthday = birthday;
        this.nativePlace = nativePlace;
    }


    public void setStuId(String StuId) {
        this.StuId = StuId;
    }

    public String getStuId() {
        return StuId;
    }

    public void setStuName(String StuName) {
        this.StuName = StuName;
    }

    public String getStuName() {
        return StuName;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public String getGender() {
        return gender;
    }

    public void setBirthday(String birthday) {
        this.birthday = birthday;
    }

    public String getBirthday() {
        return birthday;
    }

    public void setNativePlace(String nativePlace) {
        this.nativePlace = nativePlace;
    }

    public String getNativePlace() {
        return nativePlace;
    }

    @Override
    public void input() {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter student ID: ");
        StuId = sc.nextLine();
        System.out.print("Enter student name: ");
        StuName = sc.nextLine();
        System.out.print("Enter gender: ");
        gender = sc.nextLine();
        System.out.print("Enter birthday: ");
        birthday = sc.nextLine();
        System.out.print("Enter native place: ");
        nativePlace = sc.nextLine();
    }

    @Override
    public void display() {
        System.out.println("Student ID: " + StuId);
        System.out.println("Student name: " + StuName);
        System.out.println("Gender: " + gender);
        System.out.println("Birthday: " + birthday);
        System.out.println("Native place: " + nativePlace);
    }
}


public class StudentMark implements ic {
    private String StuId;
    private String className;
    private String subjectName;
    private int semester;
    private float mark;

    public StudentMark() {
    }

    public StudentMark(String StuId, String className, String subjectName, int semester, float mark) {
        this.StuId = StuId;
        this.className = className;
        this.subjectName = subjectName;
        this.semester = semester;
        this.mark = mark;
    }


    public void setStuId(String StuId) {
        this.StuId = StuId;
    }

    public String getStuId() {
        return StuId;
    }

    public void setClassName(String className) {
        this.className = className;
    }

    public String getClassName() {
        return className;
    }

    public void setSubjectName(String subjectName) {
        this.subjectName = subjectName;
    }

    public String getSubjectName() {
        return subjectName;
    }

    public void setSemester(int semester) {
        this.semester = semester;
    }

    public int getSemester() {
        return semester;
    }

    public void setMark(float mark) {
        this.mark = mark;
    }

    public float getMark() {
        return mark;
    }

    @Override
    public void input() {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter student ID: ");
        StuId = sc.nextLine();
        System.out.print("Enter class name: ");
        className = sc.nextLine();
        System.out.print("Enter subject name: ");
        subjectName = sc.nextLine();
        System.out.print("Enter semester: ");
        semester = sc.nextInt();
        System.out.print("Enter mark: ");
        mark = sc.nextFloat();
        sc.nextLine(); 
    }

    @Override
    public void display() {
        System.out.println("Student ID: " + StuId);
        System.out.println("Class name: " + className);
        System.out.println("Subject name: " + subjectName);
        System.out.println("Semester: " + semester);
        System.out.println("Mark: " + mark);
    }
}

package aptech.mark;

public class StudentMarkTotal extends StudentMark {
    private int totalExamSubject;
    private float averageMark;

    public StudentMarkTotal() {
    }

    public StudentMarkTotal(String StuId, String className, String subjectName, int semester, float mark, int totalExamSubject, float averageMark) {
        super(StuId, className, subjectName, semester, mark);
        this.totalExamSubject = totalExamSubject;
        this.averageMark = averageMark;
    }

    // Setters and getters
    public void setTotalExamSubject(int totalExamSubject) {
        this.totalExamSubject = totalExamSubject;
    }

    public int getTotalExamSubject() {
        return totalExamSubject;
    }

    public void setAverageMark(float averageMark) {
        this.averageMark = averageMark;
    }

    public float getAverageMark() {
        return averageMark;
    }

    public void calculateTotalExamSubject(StudentMarkTotal[] list) {
        totalExamSubject = list.length;
    }

    public void calculateAverageMark(StudentMarkTotal[] list) {
        float sum = 0;
        for (StudentMarkTotal s : list) {
            sum += s.getMark();
        }
        averageMark = sum / list.length;
    }
}


public class ex154 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int choice;
        StudentAptech[] students = null;
        StudentMarkTotal[] marks = null;

        do {
            System.out.println("1. Enter information for n Aptech students.");
            System.out.println("2. Enter m exam scores for these students.");
            System.out.println("3. Sort students by name and display.");
            System.out.println("4. Find exam score information by student id.");
            System.out.println("5. Exit");
            System.out.print("Choose an option: ");
            choice = sc.nextInt();
            sc.nextLine(); // Consume newline left over

            switch (choice) {
                case 1:
                    System.out.print("Enter the number of students: ");
                    int n = sc.nextInt();
                    sc.nextLine(); // Consume newline left over
                    students = new StudentAptech[n];
                        for (int i = 0; i < n; i++) {
                        System.out.println("Enter information for student " + (i + 1) + ":");
                        StudentAptech student = new StudentAptech();
                        student.input();
                        students[i] = student;
                    }
                    break;

                case 2:
                    if (students == null) {
                        System.out.println("Please enter student information first.");
                        break;
                    }

                    System.out.print("Enter the number of exam scores: ");
                    int m = sc.nextInt();
                    sc.nextLine(); // Consume newline left over
                    marks = new StudentMarkTotal[m];

                    for (int i = 0; i < m; i++) {
                        System.out.println("Enter exam score information for student " + (i + 1) + ":");
                        StudentMarkTotal mark = new StudentMarkTotal();
                        mark.input();

                        boolean studentFound = false;
                        for (StudentAptech student : students) {
                            if (student.getStuId().equals(mark.getStuId())) {
                                studentFound = true;
                                break;
                            }
                        }

                        if (studentFound) {
                            marks[i] = mark;
                        } else {
                            System.out.println("Student ID not found. Please enter a valid ID.");
                            i--; // Repeat the loop iteration
                        }
                    }
                    break;

                case 3:
                    if (students == null) {
                        System.out.println("Please enter student information first.");
                        break;
                    }

                    Arrays.sort(students, Comparator.comparing(StudentAptech::getStuName));
                    for (StudentAptech student : students) {
                        student.display();
                    }
                    break;

                case 4:
                    if (marks == null) {
                        System.out.println("Please enter exam score information first.");
                        break;
                    }

                    System.out.print("Enter the student ID to search for: ");
                    String searchId = sc.nextLine();
                    boolean found = false;

                    for (StudentMarkTotal mark : marks) {
                        if (mark.getStuId().equals(searchId)) {
                            mark.display();
                            found = true;
                            break;
                        }
                    }

                    if (!found) {
                        System.out.println("Student ID not found.");
                    }
                    break;

                case 5:
                    System.out.println("Goodbye!");
                    break;

                default:
                    System.out.println("Invalid choice. Try again.");
            }
        } while (choice != 5);
    }
}



